### PR TITLE
Pyinstaller: add VERSIONINFO resource to Windows executables.

### PR DESCRIPTION
**Pull request by bitraid**
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/3857)

### CODE REVIEW

**Comment by bitraid** [comment]
The only reason i thought for keeping the file, is for the spec file to work without having to run the scripts - but i don't think that it will really be an issue for anyone. I'll make the changes soon.

File: scripts/dev/build_release.py [hunk 1]
```diff
@@ -232,6 +233,9 @@ def build_windows():
 
     artifacts = []
 
+    utils.print_title("Updating VersionInfo file")
+    gen_versioninfo
```
Comment (The-Compiler) [inline scripts/dev/build_release.py]: You're not actually calling anything here, and everything is at the toplevel of your script - that means it'll actually all be run when imported above.

Instead, the script should have a `def main():` function doing the work (everything after the imports), and a block like this at the bottom:

```python
if __name__ == '__main__':
    main()
```

then the script can still be run directly, but you can also run `gen_versioninfo.main()` here.

File: misc/qutebrowser.spec [hunk 2]
```diff
@@ -59,7 +59,8 @@ exe = EXE(pyz,
           debug=False,
           strip=False,
           upx=False,
-          console=False )
+          console=False,
+          version='misc/file_version_info.txt')
```
Comment (The-Compiler) [inline misc/qutebrowser.spec]: I'd expect to see this documented [somewhere here](https://pyinstaller.readthedocs.io/en/latest/spec-files.html), but I don't - any idea where I can find documentation about it?

File: scripts/dev/gen_versioninfo.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+import sys
+
+from PyInstaller.utils.win32 import versioninfo as vs
+
+sys.path.insert(0, os.path.join(os.path.dirname(__file__), os.pardir,
+                                os.pardir))
+
+import qutebrowser
+
+out_filename = 'misc/file_version_info.txt'
```
Comment (The-Compiler) [inline scripts/dev/gen_versioninfo.py]: Please also call `utils.change_cwd()` before writing the file to make sure it ends up at the right place - see e.g. `build_release.py` for an example.

File: scripts/dev/gen_versioninfo.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+filevers = qutebrowser.__version_info__ + (0,)
+prodvers = qutebrowser.__version_info__ + (0,)
+str_filevers = qutebrowser.__version__
+str_prodvers = qutebrowser.__version__
+
+comment_text = """\
+A keyboard-focused browser with a minimal GUI.\
+"""
```
Comment (The-Compiler) [inline scripts/dev/gen_versioninfo.py]: I think you can take this from `qutebrowser.__doc__`

File: scripts/dev/gen_versioninfo.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+copyright_text = """\
+© 2014-2018 Florian Bruhin (The Compiler) <mail@qutebrowser.org>\
+"""
```
Comment (The-Compiler) [inline scripts/dev/gen_versioninfo.py]: `qutebrowser.__copyright__`

File: scripts/dev/gen_versioninfo.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+trademark_text = """\
+qutebrowser is free software under the GNU General Public License\
+"""
```
Comment (The-Compiler) [inline scripts/dev/gen_versioninfo.py]: Why not just make this a normal one-line string?

File: scripts/dev/gen_versioninfo.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+                     vs.StringStruct('ProductName', "qutebrowser"),
+                     vs.StringStruct('ProductVersion', f'{str_prodvers}')])]),
+    vs.VarFileInfo([vs.VarStruct('Translation', [1033, 1200])])]
+
+file_version_info = vs.VSVersionInfo(ffi, kids)
+
+with open(out_filename, 'w', encoding='utf-8') as f:
+    f.write(f'{file_version_info}')
+
+f.close()
```
Comment (The-Compiler) [inline scripts/dev/gen_versioninfo.py]: No need for this - the `with` block takes care of closing automatically.

File: scripts/dev/gen_versioninfo.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+                     vs.StringStruct('LegalTrademarks', f'{trademark_text}'),
+                     vs.StringStruct('OriginalFilename', "qutebrowser.exe"),
+                     vs.StringStruct('ProductName', "qutebrowser"),
+                     vs.StringStruct('ProductVersion', f'{str_prodvers}')])]),
+    vs.VarFileInfo([vs.VarStruct('Translation', [1033, 1200])])]
+
+file_version_info = vs.VSVersionInfo(ffi, kids)
+
+with open(out_filename, 'w', encoding='utf-8') as f:
+    f.write(f'{file_version_info}')
```
Comment (The-Compiler) [inline scripts/dev/gen_versioninfo.py]: f-strings won't work on Python 3.5 which qutebrowser still supports. Instead, use `f.write(file_version_info)` directly, or possibly `f.write(str(file_version_info))`.

Same also for the f-strings above in `vs.StringTable`.

File: scripts/dev/gen_versioninfo.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+                     vs.StringStruct('CompanyName', "qutebrowser.org"),
+                     vs.StringStruct('FileDescription', "qutebrowser"),
+                     vs.StringStruct('FileVersion', f'{str_filevers}'),
+                     vs.StringStruct('InternalName', "qutebrowser"),
+                     vs.StringStruct('LegalCopyright', f'{copyright_text}'),
+                     vs.StringStruct('LegalTrademarks', f'{trademark_text}'),
+                     vs.StringStruct('OriginalFilename', "qutebrowser.exe"),
+                     vs.StringStruct('ProductName', "qutebrowser"),
+                     vs.StringStruct('ProductVersion', f'{str_prodvers}')])]),
+    vs.VarFileInfo([vs.VarStruct('Translation', [1033, 1200])])]
```
Comment (The-Compiler) [inline scripts/dev/gen_versioninfo.py]: What does this mean?

File: misc/qutebrowser.spec (hunk 2 above)
Comment (bitraid) [reply misc/qutebrowser.spec]: The documentation is [here](https://pyinstaller.readthedocs.io/en/latest/usage.html?highlight=vsversion#capturing-windows-version-data) but the usage on spec files is missing. Command line usage is [here](https://pyinstaller.readthedocs.io/en/latest/usage.html?highlight=vsversion#windows-specific-options)

File: scripts/dev/gen_versioninfo.py (hunk 9 above)
Comment (bitraid) [reply scripts/dev/gen_versioninfo.py]: It's the identifiers for the language (1033=English) and the character set (1200=unicode). See [here](https://msdn.microsoft.com/en-us/library/windows/desktop/aa381057(v=vs.85).aspx)

**Comment by The-Compiler** [comment]
Thanks! I tried adding some documentation for the "magic numbers" in the script in 8949afc2bfc06460c4325140035c686dc24e0613 - I hope those reflect reality somewhat... I sure find Windows stuff confusing, so thanks for helping out there! :wink:

**Pull request merged**